Composite: remove store prop and useCompositeStore hook

ciampo opened the pull request:
<!-- Thanks for contributing to Gutenberg! Please follow the Gutenberg Contributing Guidelines:
https://github.com/WordPress/gutenberg/blob/trunk/CONTRIBUTING.md -->

## What?
<!-- In a few words, what is the PR actually doing? -->

Remove the `useCompositeStore` export and the `store` prop from the `Composite` component

## Why?
<!-- Why is this PR necessary? What problem is it solving? Reference any existing previous issue(s) or PR(s), but please add a short summary here, too -->

See https://github.com/WordPress/gutenberg/issues/63704#issuecomment-2305291168

## How?
<!-- How is your PR addressing the issue at hand? What are the implementation details? -->

- Removed the `useCompositeStore` from the component's exported APIs;
- Removed the `store` prop from the `Composite` component;
- Obfuscated the type of the `store` prop from `Composite.Context`;
- Updated README and JSDocs;
- Updated Storybook.

All usages of the `useCompositeStore` hook and the `store` prop were removed in separate PRs:

- https://github.com/WordPress/gutenberg/pull/64832 (requirement to all subsequent PRs)
- https://github.com/WordPress/gutenberg/pull/64850
- https://github.com/WordPress/gutenberg/pull/64833
- https://github.com/WordPress/gutenberg/pull/64983
- https://github.com/WordPress/gutenberg/pull/64985
- https://github.com/WordPress/gutenberg/pull/64987
- https://github.com/WordPress/gutenberg/pull/65038
- https://github.com/WordPress/gutenberg/pull/65039
- https://github.com/WordPress/gutenberg/pull/65040
- https://github.com/WordPress/gutenberg/pull/65041
- https://github.com/WordPress/gutenberg/pull/65042
- https://github.com/WordPress/gutenberg/pull/65043
- https://github.com/WordPress/gutenberg/pull/65044
- https://github.com/WordPress/gutenberg/pull/65047


## Testing Instructions
<!-- Please include step by step instructions on how to test this PR. -->
<!-- 1. Open a post or page. -->
<!-- 2. Insert a heading block. -->
<!-- 3. etc. -->

No runtime changes, apart from [this one](https://github.com/WordPress/gutenberg/pull/64723#discussion_r1744026078).

- Project builds and all tests pass;
- Make sure there are no usages of `useCompositeStore` left, apart from the `Composite` component itself;
- Make sure there are no instances left of passing the `store` prop to `Composite` or its sub-components;
- Make sure `Composite`'s Storybook docs are up to date and the examples work as expected;
- Make sure all JSDocs and README info is up to date (ie. doesn't recommend using `useCompositeStore` anymore)

ciampo committed:
Composite: remove store prop and useCompositeStore hook

Instead, expose props directly on top-level component

ciampo committed:
Add tsx extension to context

ciampo committed:
Update usage examples

ciampo committed:
Update legacy implementation

ciampo committed:
Remove from private APIs

ciampo committed:
Update and simplify Storybook

ciampo commented on packages/block-editor/src/components/block-patterns-list/index.js:
```diff
@@ -206,43 +226,38 @@ function BlockPatternsList(
-	const compositeStore = useCompositeStore( { orientation } );
-	const { setActiveId } = compositeStore;
-
-	useEffect( () => {
-		// We reset the active composite item whenever the
-		// available patterns change, to make sure that
-		// focus is put back to the start.
-		setActiveId( undefined );
-	}, [ setActiveId, shownPatterns, blockPatterns ] );
```
Moved this effect to a child of `Composite` (called `BlockPatternsListContentWrapper`), which can access the `store` via content.

[TODO: add video of the functionality]

> [!NOTE]
> Some interactions seem to be a bit broken also on trunk: in Chrome, when loading the patterns modal, I see a warning about `aria-hidden` not being applied to a tabbable element. When clicking a pattern from the list, focus seems to go leave the modal, despite the modal staying open.

tyxla replied:
Yeah, let's take note of those and open issues so they can be resolved separately.

ciampo commented on packages/components/src/alignment-matrix-control/index.tsx:
```diff
@@ -58,25 +57,26 @@ export function AlignmentMatrixControl( {
-		rtl: isRTL(),
-	} );
-
-	const activeId = useStoreState( compositeStore, 'activeId' );
```
Using `data-active-item` attribute to style active grid point, instead of using JS

ciampo commented on packages/components/src/circular-option-picker/circular-option-picker.tsx:
```diff
@@ -86,23 +86,22 @@ function ListboxCircularOptionPicker(
-
-	const compositeContext = {
-		baseId,
-		compositeStore,
```
Instead of using the store, we're explicitly controlling the `Composite` component and letting the individual option items set the active id (via react context), should they need to.

ciampo commented on packages/components/src/composite/stories/index.story.tsx:
```diff
@@ -91,116 +88,112 @@ const meta: Meta< typeof UseCompositeStorePlaceholder > = {
 		},
 	},
 };
+
+// TODO: example accessing the store
+
+// TODO: example across SlotFill forwarding context
```
I will add these Storybook examples if we thing that the proposed approach makes sense

ciampo commented on packages/components/src/composite/types.ts:
```diff
@@ -6,13 +6,13 @@ import type * as Ariakit from '@ariakit/react';
 export type CompositeContextProps =
```
TODO: consider changing the context type to something like

```ts
export type CompositeContextProps = {
	/**
	 * The component store, used for advanced usage of the component.
	 */
	store?: Ariakit.CompositeStore;
};
```

for easier prop destructuring when consuming it

ciampo commented on packages/dataviews/src/components/dataviews-filters/search-widget.tsx:
```diff
@@ -108,85 +144,94 @@ function ListBox( { view, filter, onChangeView }: SearchWidgetProps ) {
 				filter.name
 			) }
 			onFocusVisible={ () => {
+				// `onFocusVisible` needs the `Composite` component to be focusable,
+				// which is implicitly achieved via the `virtualFocus: true` option
+				// in the `useCompositeStore` hook.
+				const compositeStore = compositeStoreRef.current;
+				if ( ! compositeStore ) {
+					return;
+				}
```
Since we need to access the store inside a prop for the top-level component, I created a little helper component that reads the store from the `Composite.Context` and passes it back to its parent via a callback.

It is a bit convoluted, but it's kind of a necessity caused by the fact that we're not defining a store outside of `Composite`, nor we're using a `Composite.Provider` that wraps around `Composite`.

It's not pretty, but it seems to do the trick — and hopefully it seems like an edge case. Although I'd be curious to hear what folks think about it, and whether they can come up with good alternatives.

tyxla replied:
It's a bit of a bummer we still kind of expose the store here. Can't we access specific parts of the store as needed, instead of the entire store?

ciampo replied:
Even if we did, it wouldn't take way from the fact that, in order to access them via context, you'd need to be a child of `Composite`, and then somehow pass them back up the react tree

ciampo commented on packages/dataviews/src/dataviews-layouts/list/index.tsx:
```diff
@@ -359,29 +386,6 @@ export default function ViewList< Item >( props: ViewListProps< Item > ) {
-		defaultActiveId: getItemDomId( selectedItem ),
-	} ) as CompositeStore; // TODO, remove once composite APIs are public
-
-	// Manage focused item, when the active one is removed from the list.
-	const isActiveIdInList = useStoreState(
-		store,
-		( state: { items: any[]; activeId: any } ) =>
-			state.items.some(
-				( item: { id: any } ) => item.id === state.activeId
-			)
-	);
-	useEffect( () => {
-		if ( ! isActiveIdInList ) {
-			// Prefer going down, except if there is no item below (last item), then go up (last item in list).
-			if ( store.down() ) {
-				store.move( store.down() );
-			} else if ( store.up() ) {
-				store.move( store.up() );
-			}
-		}
-	}, [ isActiveIdInList ] );
```
This logic was moved to a new `ViewListWrapper` component, that can access the `store` via `Composite.Context`

ciampo commented:
@WordPress/gutenberg-components can you take a look at the proposed approach? If we agree on moving forward, I'll tackle the missing changes and make it ready to review properly.

tyxla commented on packages/block-directory/src/components/downloadable-blocks-list/index.js:
```diff
@@ -27,7 +25,6 @@ function DownloadableBlocksList( { items, onHover = noop, onSelect } ) {
 
 	return (
 		<Composite
-			store={ composite }
```
It's definitely nicer than before to be rid of what looks like an arbitrary prop. Arbitrary from the PoV of the regular component consumer, of course.

tyxla commented on packages/block-editor/src/components/inserter-listbox/index.js:
```diff
@@ -12,17 +12,11 @@ export { default as InserterListboxGroup } from './group';
-
 	return (
-		<Composite store={ store } render={ <></> }>
+		<Composite focusShift focusWrap="horizontal" render={ <></> }>
```
👍 
You can see how we're more intentional here and I could argue this is more readable/comprehensible than having to search for the `store` definition and understanding the lower-level `useCompositeStore` API usage.

tyxla commented on packages/components/src/alignment-matrix-control/cell.tsx:
```diff
@@ -18,12 +24,15 @@ import type { WordPressComponentProps } from '../context';
 
+	const compositeContext = useContext( Composite.Context );
+
+	const activeId = useStoreState( compositeContext?.store, 'activeId' );
```
What do you suggest we do with `useStoreState`? Seems like to allow for such patterns outside of the package, we'll need to expose it.

ciampo replied:
My ideal outcome would be to make it so that consumers of `Composite` don't need to use the `store` at all, but let's see what where we can get.

tyxla replied:
That would obviously be perfect 🤌

tyxla commented on packages/components/src/circular-option-picker/circular-option-picker-option.tsx:
```diff
@@ -97,10 +96,10 @@ export function Option( {
 		<OptionAsOption
 			{ ...commonProps }
-			compositeStore={ compositeStore }
+			compositeStore={ compositeContext?.store }
```
Should we also migrate this `compositeStore` prop to actually pass specific props like `activeId` and `setActiveId` instead? This will keep it more consistent with the rest of the approach being proposed here.

tyxla commented on packages/components/src/composite/stories/utils.tsx:
```diff
@@ -1,76 +0,0 @@
-// The output generated by Storybook for these components is
-// messy, so we apply this transform to make it more useful
-// for anyone reading the docs.
-export function transform( code: string, context: StoryContext ) {
```
❤️ Cool cleanup here!

tyxla commented on packages/dataviews/src/dataviews-layouts/list/index.tsx:
```diff
@@ -319,6 +317,35 @@ function ListItem< Item >( {
+function ViewListWrapper( {
+	children,
+}: {
+	children: ( store?: Ariakit.CompositeStore ) => React.ReactNode;
```
If we really want to conceal the store, we might need to come up with some APIs that allow us to work with the store state directly. That might help simplify use cases like this, and allow us to refrain from going so low-level.

tyxla reviewed (commented):
The approach seems very sensible for the most part 👍 

It appears we might need to expose some additional utilities to work with parts of the store, instead of the entire store as is currently necessary in dataviews. 

If we continue to work directly with the store in dataviews, it feels like we're defeating the purpose of concealing the store prop in the first place. Would be nice if we can explore some alternatives, and expose only parts of the store that we'll actually need. 

Would love to hear feedback from @mirka as well.

ciampo commented:
Ideally, I'd like to find ways for consumers of the component to achieve the desired behaviour without the need for the `store` (or any store state / store methods). This could be mostly achieved by using the `Composite` component in controlled mode specifically to certain props (like `activeId`) — after all, that's how other ariakit-based components work too.

For example:

- [need to reset the active id](https://github.com/WordPress/gutenberg/pull/64723#discussion_r1727757331)? Instead of calling `store.setActiveId()`, just control the `activeId` prop on `Composite`
- [need to focus the first composite item](https://github.com/WordPress/gutenberg/pull/64723#discussion_r1727771015)? You can grab a ref to it and explicitly call `elementRef.current.focus()` and, if needed, set the `activeId` prop on `Composite`
- [need to select a new item when the current one is removed from the list](https://github.com/WordPress/gutenberg/pull/64723/files#r1728790178)? Hopefully, that could also be done with some custom logic + controlling the `activeId` prop

I'll see if I can apply these ideas into practice in this PR

ciampo commented:
I pushed a few commits where I refactored  some of the the most involved usages of the composite store — they now rely on controlling the `activeId` state instead.

There are still some usages left (and one usage to polish), but I'd say that this is looking promising and I think that we may be able to move away from using `store` outside of the `Composite` component altogether. If that's the case, I'm also thinking of changing the `Composite.Context` types so that there's no "public" mention of a `store`.

ciampo commented on packages/components/src/circular-option-picker/circular-option-picker-option.tsx:
```diff
@@ -46,18 +45,18 @@ function UnforwardedOptionAsOption(
+
+	useEffect( () => {
+		if ( isSelected && ! activeId ) {
+			window.setTimeout( () => setActiveId?.( id ), 0 );
```
Without this `setTimeout`, the `setActiveId` call would be immediately overridden by Composite's internal logic that picks the first item as the active item (if no initial active item is passed). Not sure if there's a better way.

tyxla replied:
`window.queueMicrotask` should do? It's not necessarily better, but maybe a bit more intentional.

ciampo replied:
I can try that in https://github.com/WordPress/gutenberg/pull/64833 if we think we should move on with the approach

mirka replied:
Do we need a test for this?

I fiddled a bit and found that having access to `setItems` will avoid the need for a timeout. Despite the additional lines of code, I _think_ this is preferable in terms of what we can recommend to extenders. "Use a timeout/rAF/queueMicrotask in this specific scenario" is way too random and imperative for us to write in docs 😅 Also, I think we will encounter use cases where we need access to `items` anyway, in lieu of store methods like `first()` and `last()`.

```diff
diff --git a/packages/components/src/circular-option-picker/circular-option-picker-option.tsx b/packages/components/src/circular-option-picker/circular-option-picker-option.tsx
index 1136f659b1..fb251f8c53 100644
--- a/packages/components/src/circular-option-picker/circular-option-picker-option.tsx
+++ b/packages/components/src/circular-option-picker/circular-option-picker-option.tsx
@@ -54,7 +54,7 @@ function UnforwardedOptionAsOption(
 
 	useEffect( () => {
 		if ( isSelected && ! activeId ) {
-			window.setTimeout( () => setActiveId?.( id ), 0 );
+			setActiveId?.( id );
 		}
 	}, [ isSelected, setActiveId, activeId, id ] );
 
diff --git a/packages/components/src/circular-option-picker/circular-option-picker.tsx b/packages/components/src/circular-option-picker/circular-option-picker.tsx
index adf2b386e5..55f17903a7 100644
--- a/packages/components/src/circular-option-picker/circular-option-picker.tsx
+++ b/packages/components/src/circular-option-picker/circular-option-picker.tsx
@@ -8,7 +8,7 @@ import clsx from 'clsx';
  */
 import { useInstanceId } from '@wordpress/compose';
 import { isRTL } from '@wordpress/i18n';
-import { useMemo, useState } from '@wordpress/element';
+import { useEffect, useMemo, useState } from '@wordpress/element';
 
 /**
  * Internal dependencies
@@ -89,6 +89,7 @@ function ListboxCircularOptionPicker(
 	const [ activeId, setActiveId ] = useState< string | null | undefined >(
 		undefined
 	);
+	const [ items, setItems ] = useState();
 
 	const contextValue = useMemo(
 		() => ( {
@@ -99,6 +100,12 @@ function ListboxCircularOptionPicker(
 		[ baseId, activeId, setActiveId ]
 	);
 
+	useEffect( () => {
+		if ( items && ! activeId ) {
+			setActiveId( items[ 0 ].id );
+		}
+	}, [ activeId, items ] );
+
 	return (
 		<div className={ className }>
 			<CircularOptionPickerContext.Provider value={ contextValue }>
@@ -110,6 +117,8 @@ function ListboxCircularOptionPicker(
 					role="listbox"
 					activeId={ activeId }
 					setActiveId={ setActiveId }
+					defaultActiveId={ null }
+					setItems={ setItems }
 				>
 					{ options }
 				</Composite>
diff --git a/packages/components/src/composite/index.tsx b/packages/components/src/composite/index.tsx
index e8e3a5811a..ce8c0575e3 100644
--- a/packages/components/src/composite/index.tsx
+++ b/packages/components/src/composite/index.tsx
@@ -56,6 +56,7 @@ export const Composite = Object.assign(
 			activeId,
 			defaultActiveId,
 			setActiveId,
+			setItems,
 			focusLoop = false,
 			focusWrap = false,
 			focusShift = false,
@@ -81,6 +82,7 @@ export const Composite = Object.assign(
 			focusShift,
 			virtualFocus,
 			orientation,
+			setItems,
 			rtl,
 		} );


```
(☝️ Rough sketch code, isn't typed and doesn't filter disabled items)

Tangent: We have this kind of `activeId` syncing logic in Tabs, too, which involves some `rAF` shenanigans. Maybe there is some reworking that can be done to avoid that as well 🤔 But I'm more forgiving of that because that's internal code and it's not like we'd need to tell extenders to do it.

ciampo replied:
Good point. Maybe I'll ping Diego in https://github.com/WordPress/gutenberg/pull/64833 and ask him what is, in his  opinion, a better approach to using timeouts/rAFs/queueMicrotasks

If we need to expose `setItems/items/defaultItems`, we can do that.

ciampo committed:
Obfuscate store type in Composite context

ciampo commented:
@tyxla and @mirka , I've refactored all usages of composite's store so that they don't need composite store anymore.

Please take a look and, if you think that the approach is good, I'll start splitting each major refactor to a separate PR.

ciampo commented:
@WordPress/gutenberg-components I'd appreciate some feedback on this approach, as it's currently blocking any further progress on stabilizing `Composite`

mirka commented on packages/dataviews/src/components/dataviews-filters/search-widget.tsx:
```diff
@@ -84,22 +91,38 @@ const getNewValue = (
+		// since the focus is on the operators control.
+		filter.operators?.length === 1 ? undefined : null
+	);
+	const firstCompositeElementRef = useRef< HTMLElement | null >( null );
```
What is this ref for?

ciampo replied:
It can be deleted. Initially I was calling `.focus()` on it, but then I switched to generating custom composite IDs , which allows to call 	`setActiveItemID` without imperatively calling `firstCompositeElementRef.focus()`. I will clean up in the dedicated PR

mirka commented on packages/dataviews/src/components/dataviews-filters/search-widget.tsx:
```diff
@@ -108,18 +131,24 @@ function ListBox( { view, filter, onChangeView }: SearchWidgetProps ) {
 			onFocusVisible={ () => {
-				if ( ! compositeStore.getState().activeId ) {
-					compositeStore.move( compositeStore.first() );
+				// `onFocusVisible` needs the `Composite` component to be focusable,
+				// which is implicitly achieved via the `virtualFocus: true` option
+				// in the `useCompositeStore` hook.
```
Is this comment outdated?

ciampo replied:
Still valuable. It explains why `onFocusVisible` is actually working, because from the prop docs one may assume it won't fire since the `focusable` prop is not set

mirka commented on packages/dataviews/src/components/dataviews-filters/search-widget.tsx:
```diff
@@ -108,18 +131,24 @@ function ListBox( { view, filter, onChangeView }: SearchWidgetProps ) {
 					key={ element.value }
 					render={
 						<CompositeItem
+							id={ computeCompositeItemId( baseId, element ) }
```
Is this "computing" the id in the sense that Ariakit uses some kind of id scheme and we're computing what our id would be according to that scheme? Or are we just building an arbitrary id that is consistent within our code? To me, the function name is implying the former.

ciampo replied:
The latter.

You definitely have a point, I can rename it to `generateCompositeItemId`

mirka reviewed (commented):
I'm assuming we're doing more in-depth reviews in the partial PRs later, but at a high-level this certainly looks viable. I'm still worried about when we'll need any of the store methods, but it seems ok for now?

ciampo commented:
Thank you, @mirka  — I'll start splitting the more involved reactors in separate PRs, we can take it from there

tyxla commented:
Thanks for working on it @ciampo 👍 the approach makes sense to me as well. Will take a look at the more focused PRs.

ciampo committed:
Remove store prop from Composite

ciampo committed:
Update README

ciampo committed:
Update comment

ciampo committed:
Add Context subcomponent to Storybook

ciampo commented on packages/components/src/composite/index.tsx:
```diff
@@ -88,8 +84,6 @@ export const Composite = Object.assign(
 			rtl,
 		} );
 
-		const store = storeProp || newStore;
```
This is the main (only) runtime change — we don't accept a `store` prop anymore on `Composite`. Although it shouldn't really make a difference, since we've already migrated all usages that were passing a `store` prop.

ciampo committed:
Inherit default story args in each story

ciampo commented on packages/components/src/composite/stories/index.story.tsx:
Storybook changes:

- Moved from `StoryFn` to `StoryObj` style for Storybook examples;
- Changed the main component from `UseCompositeStorePlaceholder` to `Composite` 🎉 
- Removed the Storybook utils, which are not needed anymore
- Added `Composite.Context` as a subcomponent

mirka replied:
Love this CSF 3 refactor ❤️

ciampo commented on packages/components/src/slot-fill/index.tsx:
```diff
@@ -71,6 +71,7 @@ export function Provider( {
 		</SlotFillProvider>
 	);
 }
+Provider.displayName = 'SlotFillProvider';
```
For better Storybook code snippets and debugging with React DevTools

ciampo commented on packages/dataviews/src/components/dataviews-filters/search-widget.tsx:
```diff
@@ -126,8 +126,7 @@ function ListBox( { view, filter, onChangeView }: SearchWidgetProps ) {
 				// `onFocusVisible` needs the `Composite` component to be focusable,
-				// which is implicitly achieved via the `virtualFocus: true` option
-				// in the `useCompositeStore` hook.
+				// which is implicitly achieved via the `virtualFocus` prop.
```
This comment is out of date after the recent refactor

ciampo commented on packages/components/src/composite/context.tsx:
Using `.tsx` instead of `.ts` is more Storybook-friendly for types extraction.

ciampo committed:
CHANGELOG

ciampo commented:
After extracting every refactor to a separate PR, all of those PRs are finally merged. I rebased and updated the PR description.

@tyxla @mirka this PR is finally ready for a final review, and only includes changes to the `Composite` component itself.

ciampo committed:
Add SlotFillProvider to Storybook code snippet

mirka commented on packages/components/src/composite/types.ts:
```diff
@@ -3,16 +3,14 @@
+	/**
+	 * The component store, used for advanced usage of the component.
+	 */
+	store?: unknown;
```
Sorry if this was already discussed somewhere — why is this `unknown` now? Might need a code comment since it was non-obvious to me.

ciampo replied:
I think it was discussed in one of the (many) inline comments, definitely quite easy to miss 😅 

We thought that, in order to discourage `store` usage, we'd change the type to `unknown` — which still allows TypeScript to correctly validate scenarios like context forwarding (check the "With Slot Fill" example), but it wouldn't give insight into what `store` is and does.

I [added a comment](https://github.com/WordPress/gutenberg/pull/64723/commits/3bd552c4d0406f59d4ae62bfbfb9aff037122fc4) for now to clarify.

What do you think? Too much?

Happy to make changes in a follow-up.

mirka replied:
Ok, I see 😂

mirka commented on packages/components/src/composite/stories/index.story.tsx:
```diff
@@ -93,116 +92,112 @@ const meta: Meta< typeof UseCompositeStorePlaceholder > = {
-	);
+export const Default: StoryObj< typeof Composite > = {
+	args: {
+		rtl: isRTL(),
```
Mmm, I'm having some thoughts about this `rtl` feature.

One, I see we're still referring to `store` in the description here so we should rewrite that.

https://github.com/WordPress/gutenberg/blob/955ac5de18decdb5eaec7aa394b9771437ced14f/packages/components/src/composite/types.ts#L117-L127

Two, `rtl: isRTL()` as part of the args object is not going to work dynamically with the Storybook's RTL switcher, unlike when it was executed as part of the story function.

Three, this then raises the question of whether we should have `rtl={ isRTL() }` be the default behavior for the component. I can imagine a lot of consumers forgetting to do this, and they can override it if they need to. What do you think? At the very least I think this practice should be featured more heavily in the docs.

ciampo replied:
> we're still referring to `store` in the description here so we should rewrite that.

👍 

> `rtl: isRTL()` as part of the args object is not going to work dynamically with the Storybook's RTL switcher

Good point — it may work if Storybook does a full re-render, but it any case it's better to improve it on our end

> whether we should have `rtl={ isRTL() }` be the default behavior for the component.

I agree that this would be a healthy default.

ciampo replied:
https://github.com/WordPress/gutenberg/pull/64723/commits/12f7d1063623cab9f7f1b7382c461735f1049d28

mirka reviewed (approved):
Beautiful, great work on this! The `rtl` thing can be kicked to a follow-up, by the way.

ciampo committed:
Add extra comment to the context store

ciampo committed:
Better `rtl` default value and docs

tyxla commented:
Great to see this one land! 🚀
